### ISSUE TITLE
Spike: expo as deliver mechanism

### ISSUE THREAD

**Issue by bodymindarts** [GaloyMoney/galoy-mobile#107]
Problem:
develop a scalable way to deliver the galoy mobile app to multiple communities with their own custom branding / skins.

Possible solution:
https://expo.dev is an attractive framework for delivering react-native apps that provides OTA updates out of the box. This would enable us to add new communities to the app dynamically without needing to upload multiple apps to the App Store.
There are multiple options for how to choose the correct 'skin' for the app. The user could choose it at login, the login itself could be associated with a specific instance or a custom download link might transfer the information. The paid for delivery service from expo also makes it possible to have multiple releases of the app that are pre-skinned without significant additional complexity.

All of these things are doable with 'vanilla' react-native but it looks like moving to expo would be the best way to achieve this goal.

I propose to at least investigate how much work it would be to execute.

**Comment by cem2ran** [comment]
More context on EAS build: 
https://docs.expo.io/build/introduction/
https://blog.expo.dev/expo-application-services-eas-build-and-submit-fc1d1476aa2e

**Comment by bodymindarts** [comment]
On hold pending stabilisation of functionality that lets you add custom native dependencies.

Expo was used in the past and deemed unsatisfying given the limitations that was there at that point.

**Comment by nicolasburtey** [comment]
@bodymindarts @cem2ran our current build system (fastlane) have also support for building multiples app it seems: https://github.com/fastlane/examples/tree/master/SunApps

**Comment by cem2ran** [comment]
Fastlane will work just fine yes. What I explained @bodymindarts was what I'm currently using for white-labeling and that if I were to start with a new project I'd go with Expo even with something like Galoy that uses custom modules. Brought Expo up as an option as it now allows for the use of custom native modules as per the linked articles without sacrificing the niceties that Expo provides.

The complexity of multiple/white-labeled apps in my mind lies in figuring out how do you customize iOS and Android configurations. The two approaches I'm aware of is:

1. To have variations defined as separate iOS configuration and Android productFlavors. Here's a good article series describing this and also other useful details:
https://blog.swmansion.com/build-and-publish-multiple-rn-apps-from-the-same-codebase-part-ii-815d0fffc222

2. Another approach (which I've utilized) is to reconfigure the iOS and Android configurations/folders. I've done this indirectly via the Expo [app.json/app.config.js](https://docs.expo.io/workflow/configuration), but I think this could be done with React Native projects too, but would probably require a bit more work, unless Fastlane in fact can do this for you? Can you tell Fastlane to use a specific asset for app icon, splash screen, entitlements, firebase config...

I have a feeling that the following step in the SunApps example may be doing this: 🤷‍♂️ https://github.com/fastlane/examples/blob/e82ae4c9ef4102c25855a78f0d22c3a7854150cf/SunApps/Fastfile#L28

I would start by defining a json file or directory per app containing assets, colors, features, app store description, screenshots and configuration needed for each app. On CI I would have job per app running in parallel. This allows you to easily see if any app builds and deploys fail. For each job I would define an environment variable e.g. `app: bitcoin_beach`, that maps to the specific configuration, makes necessary changes to the iOS and Android configurations and then simply runs the Galoy fastfile with the right values substituted.

**Comment by bodymindarts** [comment]
As far as I have understood so far expo would give us the following benefits:
- better dev workflow (hot reloads)
- better rollout mechanism (OTA)
- better support for multi-app builds

OTOH the existing rollout mechanism is already setup and support for multi-app builds are possible with fastlane / circle ci.
Regardless of the long term rollout mechanism we must move to a code-base structure that enables simple configuration of theming / branding. As that is independent of expo or not this should be its own ticket.
